agent/stats/engine_integ_test.go

```diff
@@ -220,89 +191,275 @@ func TestStatsEngineWithNewContainers(t *testing.T) {
+	})
 
-	if len(taskMetrics) != 1 {
-		t.Fatalf("Incorrect number of tasks. Expected: 1, got: %d", len(taskMetrics))
+	// Wait for containers from previous tests to transition states.
+	time.Sleep(checkPointSleep)
+
+	engine.cluster = defaultCluster
+	engine.containerInstanceArn = defaultContainerInstance
+
+	err = client.StartContainer(container.ID, nil)
+	require.NoError(t, err, "starting container failed")
+	defer client.StopContainer(container.ID, defaultDockerTimeoutSeconds)
+
+	containerChangeEventStream := eventStream("TestStatsEngineWithExistingContainers")
+	taskEngine := ecsengine.NewTaskEngine(&config.Config{}, nil, nil, containerChangeEventStream, nil, dockerstate.NewTaskEngineState(), nil)
+	containers := []*api.Container{
+		{
+			Name:            "container-health",
+			HealthCheckType: "docker",
+		},
+	}
+	testTask := &api.Task{
+		Arn:                 "container-health-task",
+		DesiredStatusUnsafe: api.TaskRunning,
+		KnownStatusUnsafe:   api.TaskRunning,
+		Family:              "docker-container-health",
+		Version:             "1",
+		Containers:          containers,
 	}
+	// Populate Tasks and Container map in the engine.
+	dockerTaskEngine, _ := taskEngine.(*ecsengine.DockerTaskEngine)
+	dockerTaskEngine.State().AddTask(testTask)
+	dockerTaskEngine.State().AddContainer(
+		&api.DockerContainer{
+			DockerID:   container.ID,
+			DockerName: "container-health",
+			Container:  containers[0],
+		},
+		testTask)
+
+	// Simulate container start prior to listener initialization.
+	time.Sleep(checkPointSleep)
+	err = engine.MustInit(taskEngine, defaultCluster, defaultContainerInstance)
+	assert.NoError(t, err, "initializing stats engine failed")
+	defer engine.containerChangeEventStream.Unsubscribe(containerChangeHandler)
+
+	// Wait for the stats collection go routine to start.
+	time.Sleep(checkPointSleep)
+
+	// Verify the metrics of the container
+	metadata, taskMetrics, err := engine.GetInstanceMetrics()
+	assert.NoError(t, err, "error gettting instance metrics")
+	assert.NoError(t, validateMetricsMetadata(metadata), "error validating metadata")
+	assert.Len(t, taskMetrics, 1, "incorrect number of tasks")
+
 	taskMetric := taskMetrics[0]
-	if *taskMetric.TaskDefinitionFamily != taskDefinitionFamily {
-		t.Errorf("Excpected task definition family to be: %s, got: %s", taskDefinitionFamily, *taskMetric.TaskDefinitionFamily)
+	assert.Equal(t, aws.StringValue(taskMetric.TaskDefinitionFamily), "docker-container-health", "task definition family not expected")
+	assert.Equal(t, aws.StringValue(taskMetric.TaskDefinitionVersion), taskDefinitionVersion, "task definition family version not expected")
+	assert.NoError(t, validateContainerMetrics(taskMetric.ContainerMetrics, 1), "error validating container metrics")
+
+	// Verify the health metrics of container
+	healthMetadata, healthMetrics, err := engine.GetTaskHealthMetrics()
+	assert.NoError(t, err, "getting task health metrics failed")
+	require.Len(t, healthMetrics, 1)
```

Comment:
:+1: for @sharanyad 's comment. Let's re-use as much code as possible in `*_test.go` files so that we don't have repeating lines of code in all of our tests. Can you take another shot at refactoring this and moving commonly used assertions/methods into methods and/or switching to table based tests?